pm: device_runtime: API improvements

gmarull opened the pull request:
This PR introduces a few relevant changes:

1. pm: device_runtime: simplify error handling for get/put 

	  In case runtime PM is not enabled (or not built-in), the get/put
	  functions always return 0 (instead of -ENOTSUP/-ENOSYS). When runtime PM
	  is disabled, a device is left into active state. Similarly, when device
	  runtime PM is not built-in, it is safe to assume that a device will
	  be active when it is called. If a user implements a custom solution, it
	  is its responsability to make sure that a device is active when using
	  it. For all these reasons, the -ENOTSUP/-ENOSYS are error codes that
	  should always be ignored by devices using get/put, since in practice it
	  means that: device is active, function is a no-op. The example below
	  illustrates how error handling is simplified:
	  
	  ```c
	  /* before: safe to ignore -ENOSYS/-ENOTSUP since device is active (we
	   * can continue)
	   */
	  ret = pm_device_runtime_get(dev);
	  if ((ret < 0) && (ret != -ENOSYS) && (ret != -ENOTSUP)) {
		  return ret;
	  }
	  
	  /* now */
	  ret = pm_device_runtime_get(dev);
	  if (ret < 0) {
		  return ret;
	  }
	  ```
2. pm: device_runtime: suspend device on enable 

	  The pm_device_runtime_enable did not suspend devices, so it assumed that
	  the device was in a physically suspended state. This change makes sure
	  that device is left in a suspended state if the device is initially
	  active.

3. pm: device_runtime: add pm_device_runtime_init_suspended 

	  By default, device state is initialized to PM_DEVICE_STATE_ACTIVE. In
	  general, this makes sense because the device initialization function
	  will resume and configure a device, leaving it operational. However,
	  when device runtime PM is enabled, the device may not be resumed and the
	  init function will just enable device runtime PM. If that is the case,
	  this function can be used to set the initial device state to
	  PM_DEVICE_STATE_SUSPENDED.

As a result, the usage of the device runtime PM API has been improved in STM32 GPIO/pinmux drivers.

gmarull committed:
pm: device_runtime: simplify error handling for get/put

In case runtime PM is not enabled (or not built-in), the get/put
functions always return 0 (instead of -ENOTSUP/-ENOSYS). When runtime PM
is disabled, a device is left into active state. Similarly, when device
runtime PM is not built-in, it is safe to assume that a device will
be active when it is called. If a user implements a custom solution, it
is its responsability to make sure that a device is active when using
it. For all these reasons, the -ENOTSUP/-ENOSYS are error codes that
should always be ignored by devices using get/put, since in practice it
means that: device is active, function is a no-op. The example below
illustrates how error handling is simplified:

```c
/* before: safe to ignore -ENOSYS/-ENOTSUP since device is active (we
 * can continue)
 */
ret = pm_device_runtime_get(dev);
if ((ret < 0) && (ret != -ENOSYS) && (ret != -ENOTSUP)) {
	return ret;
}

/* now */
ret = pm_device_runtime_get(dev);
if (ret < 0) {
	return ret;
}
```

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: stm32: remove runtime PM get/put guards

The PM runtime API (get/put) provides an inline implementation when not
built-in. Such implementation always returns 1, so it is safe to not
guard PM runtime calls and check for < 0.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pinmux: stm32: remove runtime PM get/put guards

The PM runtime API (get/put) provides an inline implementation when not
built-in. Such implementation always returns 1, so it is safe to not
guard PM runtime calls and check for < 0.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: stm32: introduce gpio_stm32_configure_raw

Make gpio32_stm32_configure use runtime PM API (so that it can be used
externally without further effort). The raw version of the function (no
PM put/get) has been introduced for internal use.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: stm32: do not expose gpio_stm32_clock_request

The function is only used by the driver itself. This likely comes from
pre-runtime PM API usage.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: stm32: improve initialization

Call `gpio_stm32_clock_request` unconditionally with "true" if no
runtime PM is enabled, "false" otherwise. In case the GPIO was enabled
(e.g. by a bootloader), the device will be left in proper state when
runtime PM is enabled.

Also added pm_device_runtime_init_suspended to indicate that device is
in a suspended state (will prevent pm_device_runtime_enable to suspend
the device again).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device_runtime: add pm_device_runtime_init_suspended

By default device state is initialized to PM_DEVICE_STATE_ACTIVE. In
general, this makes sense because the device initialization function
will resume and configure a device, leaving it operational. However,
when device runtime PM is enabled, the device may not be resumed and the
init function will just enable device runtime PM. If that is the case,
this function can be used to set the initial device state to
PM_DEVICE_STATE_SUSPENDED.

Documentation has been updated to comment on this case, and example code
has been adjusted accordingly.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

erwango commented on drivers/pinmux/pinmux_stm32.c:
```diff
@@ -104,20 +104,12 @@ static int stm32_pin_configure(uint32_t pin, uint32_t func, uint32_t altf)
-#ifdef CONFIG_PM_DEVICE_RUNTIME
-	ret = pm_device_runtime_get(port_device);
-	if (ret != 0) {
+	ret = gpio_stm32_configure(port_device, STM32_PIN(pin), func, altf);
+	if (ret < 0) {
```
Same is required on pinctrl_stm32.c

gmarull replied:
adjusted too

teburd replied:
Can you return ret here instead and avoid the added branch?

gmarull replied:
done

erwango reviewed (commented):
One comment otherwise looks good

ceolin commented on include/pm/device_runtime.h:
```diff
@@ -61,9 +61,10 @@ int pm_device_runtime_disable(const struct device *dev);
+ * @retval >0 This function will return the current device usage count (> 0) in
+ * case it succeeds. The return value 1 indicates that the device has just been
+ * resumed (first user). In case device runtime PM is not enabled, this function
+ * will always return 1.
```
I don't think it should return "1" if device runtime pm is not enabled. 1 with this change means success for a get, so who calls it will assume that the device is active and that may not be the case ...

gmarull replied:
`pm_device_runtime_disable` leaves devices active. When using runtime PM the device state should be controlled by the runtime API. If a user decides to change the device state manually, it’s his responsibility to control when a device driver can be called or not. The API has its usage guidelines, if you do not follow them you should really know what you're doing.

ceolin replied:
that is not my point. I'm thinking in the path that the target was built without `CONFIG_PM_DEVICE_RUNTIME` and someone calls `pm_device_runtime_enable()` followed by `pm_device_runtime_get()`. Returning `1` may be interpreted that the device is active, but that is not true, the device may have not being initialized and not be active.

ceolin commented on include/pm/device_runtime.h:
```diff
@@ -130,13 +132,33 @@ int pm_device_runtime_put_async(const struct device *dev);
+ * This function can be used to suspend a device before enabling device runtime
+ * PM if it is physically active. This can be useful for devices that boot into
+ * an active state by default.
+ *
```
Why someone would need to call it before `pm_device_runtime_enable`  if the latter suspend the device ?
I didn't get 
```
The pm_device_runtime_enable call sets the device state to suspended,
however, it has no knowledge whether that is true or not.
```
 Has no knowledge about what ? Whether or not the device is being used ? If that is the case how this function solves the problem ? It seems that whoever is calling it has to know it, so couldn't the same premises be used `pm_device_runtime_enable` ?

ceolin replied:
Actually looking `pm_device_runtime_enable` it is inconsistent, only the first time it is called the device is suspended, so, if you do:
```
pm_device_runtime_enable();
....
pm_device_runtime_disable();
pm_device_runtime_enable();
```
you end up with an active device.

gmarull replied:
Some devices boot into low power mode (non-operational), others boot into active state. This information is not known by the PM subsystem, it is a device property. This is why `pm_device_runtime_enable` can't do this automatically: a physically suspended device can't be suspended. So the first case doesn't need any special handling: `pm_device_runtime_enable` will just set device as suspended. The second case, though, requires the device to be suspended first so that we end up in a consistent state (reality matches device state). If we don't do that the device will be left powered on, and, `pm_device_runtime_get` will try to resume a device that is in reality active.

ceolin replied:
I see the problem. Who is supposed to call `pm_device_runtime_enable` and who is supposed to call `pm_device_runtime_suspend` ? If they are the same, this argument does not make sense, and it could just be a parameter in `pm_device_runtime_enable`.

gmarull replied:
I have updated the approach to account for potential external usage of enable/disable. I think that your concerns should now be solved in that area. Can you please check again?

ceolin reviewed (changes requested):
I'm not sure if that is the best way to solve the possible inconsistency in `pm_device_runtime_enable`. Also changing the `pm_device_runtime_get` to always return 1 even when building without device runtime does not look right, since it means that the operation succeed and device is active (and that may not be the case)

gmarull committed:
pm: device_runtime: suspend device on enable

The pm_device_runtime_enable did not suspend devices, so it assumed that
the device was in a physically suspended state. This change makes sure
that device is left in a suspended state if the device is initially
active.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
@teburd @erwango @ceolin please re-review

ceolin commented on doc/guides/pm/device_runtime.rst:
```diff
@@ -142,20 +142,28 @@ special synchronization is required.
         /* enable device runtime power management */
-        pm_device_runtime_enable(dev);
-        ...
+        ret = pm_device_runtime_enable(dev);
```
enable / disable runtime should be done by the application and not by the driver, isn't ?

gmarull replied:
IMO, applications should be able to decide whether runtime PM is used for a particular device or not. Regarding initialization, though, I'm not sure deferring the responsibility to the application is a good idea in all situations. For example, assuming we have a flash being used by a file system, the application may never care about the flash device. Or, an application may never care about a bus (i2c/spi) but the sensor attached to it.

If drivers can enable runtime PM, we allow the STM32 GPIO scenario: device is never clocked until used. If application is the only part that should call enable, then the init function must leave device active.

ceolin replied:
Yep, that is what we had brainstormed, the problem in make this decision
per device is how to deal with dependencies. We may have a build option
that if it is selected, all devices supporting it have it enabled. Then
the application can disable per device if it wants, but will be application
responsability keep it consistent.

erwango replied:
If I got this discussion right:

I don't think application driven works well with device runtime. Technically this may be easier to handle. But from users perspective, it's complicated to explain:
"when the device runtime PM is enabled, devices are enabled only when there's a user, except at startup where devices are all enabled, and then up to the application to disable them one by one, when not required". 
This also raises the point of application genericity since additional treatment by the application would be required each time a new device is added.

I get the dependencies issue, but I don't think it is specific to this particular use case. It would raise anyway in other cases and would have to be dealt with some way or another.

ceolin replied:
I'm affraid I have not understood the comment. The basically idea is that the application can enable / disable this feature on a device. The device itself will handle when to suspend / resume the device calling pm_device_runtim_get / put. What is the case about new a device be added ? You meant add a device at runtime ?

ceolin commented on include/pm/device_runtime.h:
```diff
@@ -130,13 +137,33 @@ int pm_device_runtime_put_async(const struct device *dev);
+	struct pm_device *pm = dev->pm;
+
+	pm->state = PM_DEVICE_STATE_SUSPENDED;
+}
```
I don't think this function is right, we are trying to fix an inconsistency in the wrong place. My point is, if a device is physically suspended but its state is saying it is ACTIVE, it not only a runtime pm issue, because if you are not using pm runtime the system when going to sleep will look this device state and call `pm->action_cb`(pm->dev, PM_DEVICE_ACTION_SUSPEND);` for it, what is the same thing we are trying to avoid here.

Am I wrong ?

gmarull replied:
The thing is that all devices are initialized as `PM_DEVICE_STATE_ACTIVE` (assuming init hook will leave it in this state). However, if we consider the STM32 GPIO case where the device is never clocked, we need such mechanism so that `pm_device_runtime_enable` doesn't try to suspend it.

ceolin replied:
This smells workaround in the wrong place. What happens with this device when runtime pm is disabled (built without it) ? It initializes not clocked but the state is active, then when the system gets idle, this device is suspended and when the system wakes it is resumed. In this scenario you 
will have  pm->action_cb(pm->dev, PM_DEVICE_ACTION_SUSPEND) called with the device not clocked.

Anyway, I think we can live with this for now, since the problem is there and this solves at least for this use case. I just don't think this is the proper solution for the problem.

ceolin replied:
nevermind, I saw that in case it is built without device runtime pm you the device is clocked ...

ceolin commented:
I liked the changes, the only thing that I believe is not correct is the `pm_device_runtime_init_suspended`, IMHO it is workaround a problem only for device runtime pm case.

teburd reviewed (commented):
I need to mull this over a bit more, but I get the idea here and it makes some sense to defer turning anything on or initializing anything until you actually need to do so. At the same time I get what @ceolin is saying here, if the enable API was meant to be application driven, this is drastically changing that idea to be driver driven. Not a bad thing, but I need to think about that more. My previous +1 may have been a bit hasty

ceolin commented:
;;; Type C-c C-c to include your comment
I think we have a consensus about this API be application driven,
if we allow drivers to call it we won't be able to ensure a sane
behavior.

gmarull committed:
drivers: gpio: stm32: use synchronous put

The suspend operation for this device is a clock gating operation (i.e.
fast), so use synchronous put to skip unnecessary overhead.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on include/pm/device_runtime.h:
```diff
@@ -25,11 +25,21 @@ extern "C" {
  * @param dev Device instance.
+ *
+ * @retval 0 If the device runtime PM is disabled successfully.
+ * @retval -EALREADY If the device runtime PM is already enabled.
```
right, removed

gmarull commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -158,20 +157,12 @@ static int stm32_pin_configure(uint32_t pin, uint32_t func, uint32_t altf)
-#ifdef CONFIG_PM_DEVICE_RUNTIME
-	ret = pm_device_runtime_get(port_device);
+	ret = gpio_stm32_configure(port_device, STM32_PIN(pin), func, altf);
 	if (ret < 0) {
```
done, thanks (was due to previous >0 return path)

ceolin commented on include/pm/device_runtime.h:
```diff
@@ -130,13 +136,33 @@ int pm_device_runtime_put_async(const struct device *dev);
+ * general, this makes sense because the device initialization function will
+ * resume and configure a device, leaving it operational. However, when device
+ * runtime PM is enabled, the device may not be resumed and the init function
+ * will just enable device runtime PM. If that is the case, this function can be
```
This section is confuse, "init" is referencing the device initialization ? If so it should not enable device runtime there, only call this function to reflect the proper device state.

ceolin commented:
I will not block these changes, since it is an incremental improvement for the current code. My main concern is the introduction of this new API that is confuse to use, basically it requires a following call to `pm_device_runtime_enable()` and basically can only be used by the driver ...
